File: src/core/Routines/Events.cs
using System;
using core.Shared;

namespace core.Routines;

internal class RoutineCreated : AggregateEvent
{
    public RoutineCreated(Guid id, Guid aggregateId, DateTimeOffset when, string description, string name, Guid userId)
        : base(id, aggregateId, when)
    {
        Description = description;
        Name = name;
        UserId = userId;
    }

    public string Description { get; }
    public string Name { get; }
    public Guid UserId { get; }
}

internal class RoutineUpdated : AggregateEvent
{
    public RoutineUpdated(Guid id, Guid aggregateId, DateTimeOffset when, string description, string name)
        : base(id, aggregateId, when)
    {
        Description = description;
        Name = name;
    }

    public string Description { get; }
    public string Name { get; }
}

internal class RoutineStepAdded : AggregateEvent
{
    public RoutineStepAdded(Guid id, Guid aggregateId, DateTimeOffset when, string label, string url)
        : base(id, aggregateId, when)
    {
        Label = label;
        Url = url;
    }

    public string Label { get; }
    public string Url { get; }
}

internal class RoutineStepRemoved : AggregateEvent
{
    public RoutineStepRemoved(Guid id, Guid aggregateId, DateTimeOffset when, int index)
        : base(id, aggregateId, when)
    {
        Index = index;
    }

    public int Index { get; }
}

internal class RoutineStepMoved : AggregateEvent
{
    public RoutineStepMoved(Guid id, Guid aggregateId, DateTimeOffset when, int direction, int stepIndex)
        : base(id, aggregateId, when)
    {
        Direction = direction;
        StepIndex = stepIndex;
    }

    public int Direction { get; }
    public int StepIndex { get; }
}

internal class RoutineStepUpdated : AggregateEvent
{
    public RoutineStepUpdated(Guid id, Guid aggregateId, DateTimeOffset when, int index, string label, string url)
        : base(id, aggregateId, when)
    {
        Index = index;
        Label = label;
        Url = url;
    }

    public int Index { get; }
    public string Label { get; }
    public string Url { get; }
}

File: src/core2.fs/Routines/Events.fs
namespace core.Routines

open System
open core.Shared

[<AllowNullLiteral>]
type internal RoutineCreated(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, description: string, name: string, userId: Guid) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val Description = description with get
    member val Name = name with get
    member val UserId = userId with get

[<AllowNullLiteral>]
type internal RoutineUpdated(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, description: string, name: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val Description = description with get
    member val Name = name with get

[<AllowNullLiteral>]
type internal RoutineStepAdded(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, label: string, url: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val Label = label with get
    member val Url = url with get

[<AllowNullLiteral>]
type internal RoutineStepRemoved(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, index: int) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val Index = index with get

[<AllowNullLiteral>]
type internal RoutineStepMoved(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, direction: int, stepIndex: int) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val Direction = direction with get
    member val StepIndex = stepIndex with get

[<AllowNullLiteral>]
type internal RoutineStepUpdated(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, index: int, label: string, url: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val Index = index with get
    member val Label = label with get
    member val Url = url with get
